File: javascript/arrays/TwoSum.js
/**
 * Given an array of integers numbers that is already sorted in non-decreasing
 * order, find two numbers such that they add up to a specific target number.
 * Return the indices of the two numbers (1-indexed) as an integer array answer
 * of size 2, where 1 <= answer[0] < answer[1] <= numbers.length. The tests are
 * generated such that there is exactly one solution. You may not use the same
 * element twice.
 *
 * Example 1:
 * Input: numbers = [2,7,11,15], target = 9
 * Output: [1,2]
 * Explanation: The sum of 2 and 7 is 9. Therefore index1 = 1, index2 = 2.
 *
 * Constraints:
 * 2 <= numbers.length <= 3 * 10^4
 * -1000 <= numbers[i] <= 1000
 * numbers is sorted in non-decreasing order.
 * -1000 <= target <= 1000
 * The tests are generated such that there is exactly one solution.
 */

/**
 * 两指针分别从两边往中间扫描，如果两指针指向的元素之和等于要求的target，则分别对索引加1后
 * 返回，如果大于要求的target，则右指针向左移动一个位置，如果小于要求的target，则左指针向
 * 右移动一个位置
 */
var twoSum = function(numbers, target) {
    let i, j;

    for (i = 0, j = numbers.length - 1; i < j;) {
        if (numbers[i] + numbers[j] == target) {
            break;
        } else if (numbers[i] + numbers[j] < target) {
            i++;
        } else {
            j--;
        }
    }

    return [i + 1, j + 1];
}

let numbers = [2,7,11,15];
let target = 9;
console.log('input array: ' + numbers);
console.log('target: ' + target);
console.log('output array: ' + twoSum(numbers, target));

File: java/hashTable/TwoSum.java
/**
 * Given an array of integers nums and an integer target, return indices of the 
 * two numbers such that they add up to target. You may assume that each input would 
 * have exactly one solution, and you may not use the same element twice. You can 
 * return the answer in any order.
 * 
 * Example 1:
 * Input: nums = [2,7,11,15], target = 9
 * Output: [0,1]
 * Output: Because nums[0] + nums[1] == 9, we return [0, 1].
 * 
 * Constraints:
 * 2 <= nums.length <= 10^4
 * -10^9 <= nums[i] <= 10^9
 * -10^9 <= target <= 10^9
 * Only one valid answer exists.
 */

package hashTable;

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

public class TwoSum {

    public static void main(String[] args) {
        int[] nums = {3,2,4};
        int target = 6;
        TwoSum t = new TwoSum();
        System.out.println(Arrays.toString(t.twoSum(nums, target)));
    }
    
    /**
     * 遍历数组，每次遇到一个元素就保存目标值和元素值之差以及对应索引，当目标值和元素值之差等于
     * 某个元素的时候则返回true
     */
    public int[] twoSum(int[] nums, int target) {
        Map<Integer, Integer> map = new HashMap<Integer, Integer>();
        for (int i = 0; i < nums.length; i++) {
            if (map.containsKey(nums[i])) {
                return new int[]{map.get(nums[i]), i};
            }
            map.put(target - nums[i], i);
        }
        return new int[]{-1,-1};
    }

}
